File: python/copyLinkedList.py
class Node:
    def __init__(self, val=0, next=None, random=None) -> None:
        self.val = val
        self.next = next
        self.random = random


# Time -> 0(N) Space -> 0(N)
def copyNode(head: Node | None) -> Node | None:
    nodeMap = {}

    start = None
    newHead = None
    actualHead = head

    while head is not None:
        if start is None:
            start = Node()
            newHead = start

        start.val = head.val
        nodeMap[head] = start

        if head.next is not None:
            start.next = Node()

        head = head.next
        start = start.next

    head = newHead
    while actualHead is not None and newHead is not None:
        newHead.random = nodeMap.get(actualHead.random, None)
        actualHead = actualHead.next
        newHead = newHead.next

    return head

File: go/CopyLinkedList/CopyLinkedList.go
package main

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

// Time -> 0(N) Space -> 0(N)
func copyNode(head *Node) *Node {
	nodeMap := make(map[*Node]*Node, 0)

	var newHead, start *Node
	actualHead := head

	for head != nil {
		if start == nil {
			start = &Node{}
			newHead = start
		}

		start.Val = head.Val
		nodeMap[head] = start

		if head.Next != nil {
			start.Next = &Node{}
		}

		head = head.Next
		start = start.Next
	}

	head = newHead

	for newHead != nil && actualHead != nil {
		newHead.Random = nodeMap[actualHead.Random]
		newHead = newHead.Next
		actualHead = actualHead.Next
	}

	return head
}
